JSON-Grammar-spec.md

```diff
@@ -0,0 +1,794 @@
+
+#### Basic AST Expressions
+
+**Skip Node:**
+```json
+{"ast": null}
+```
+
+**Use Child Node:**
+```json
+{"ast": ["$", "/children/0"]}
+```
+
+**Custom Object:**
+```json
+{
+  "ast": {
+    "type": "CustomType",
+    "value": ["$", "/raw"]
+  }
+}
+```
+
+#### JSON Expression Operations
+
+**Value Extraction:**
+```json
+["$", "/raw"]              // Get raw matched text
+["$", "/children/0"]       // Get first child's AST
+["$", "/pos"]              // Get start position
+["$", "/end"]              // Get end position
+```
+
+**Type Conversion:**
+```json
+["num", ["$", "/raw"]]     // Convert to number
+["bool", ["$", "/raw"]]    // Convert to boolean
+```
+
+**String Operations:**
+```json
+["substr", ["$", "/raw"], 1, -1]  // Remove first and last character
+["len", ["$", "/raw"]]            // Get string length
+```
+
+**Array Operations:**
+```json
+["push", [[]], ["$", "/children/0"]]    // Create array with element
+["concat", ["$", "/children/0"], ["$", "/children/1"]]  // Concatenate arrays
+```
+
+**Conditional Logic:**
+```json
+["?", ["==", ["$", "/raw"], "true"], true, false]  // Ternary operator
+```
+
+**Object Construction:**
+```json
+["o.set", ["$", ""], "key", ["$", "/children/0"]]  // Set object property
+```
+
+### Children Mapping
+
+Use the `children` property to map CST child indices to AST properties:
+
+```json
+{
+  "p": [{"r": "Key"}, ":", {"r": "Value"}],
+  "children": {
+    "0": "key",
+    "2": "value"
+  }
+}
+```
+
+This creates an AST node with `key` and `value` properties instead of a `children` array.
+
+### Complex AST Examples
+
+#### Boolean Conversion
+```json
+{
+  "cst": {
+    "Boolean": {"t": ["true", "false"]}
+  },
+  "ast": {
+    "Boolean": ["==", ["$", "/raw"], "true"]
+  }
+}
+```
+
+#### Number Conversion
+```json
+{
+  "cst": {
+    "Number": "/\\d+/"
+  },
+  "ast": {
+    "Number": ["num", ["$", "/raw"]]
+  }
+}
+```
+
+#### String Unescaping
+```json
+{
+  "cst": {
+    "String": "/\"[^\"]*\"/"
+  },
+  "ast": {
+    "String": ["substr", ["$", "/raw"], 1, -1]
+  }
+}
+```
+
+#### Array Collection
+```json
+{
+  "cst": {
+    "Array": ["[", {"r": "Elements"}, "]"]
+  },
+  "ast": {
+    "Array": ["$", "/children/1"]
+  }
+}
+```
+
+#### List Flattening
+```json
+{
+  "cst": {
+    "CommaSeparated": {
+      "p": [
+        {"r": "Item"},
+        {
+          "l": {
+            "p": [",", {"r": "Item"}],
+            "ast": ["$", "/children/1"]
+          }
+        }
+      ],
+      "ast": ["concat", ["push", [[]], ["$", "/children/0"]], ["$", "/children/1"]]
+    }
+  }
+}
+```
+
+## Complete Examples
+
+### Simple Expression Parser
+
+A basic arithmetic expression parser:
+
+```json
+{
+  "start": "Expression",
+  "cst": {
+    "Expression": {"r": "Term"},
+    "Term": {"r": "Number"},
+    "Number": {
+      "t": "/\\d+/",
+      "type": "Number"
+    }
+  },
+  "ast": {
+    "Number": {
+      "type": "Number",
+      "value": ["num", ["$", "/raw"]]
+    }
+  }
+}
+```
```

Comment:
Not sure if I like this example. It mentions "arithmetic", but there is no arithmetic, it only parses a number.